pkg/resources/kafka/kafka.go

```diff
@@ -968,32 +996,50 @@ func getServiceFromExternalListener(client client.Client, cluster *v1beta1.Kafka
 	return foundLBService, nil
 }
 
-func (r *Reconciler) reorderBrokers(log logr.Logger, brokers []v1beta1.Broker) []v1beta1.Broker {
-	kClient, close, err := r.kafkaClientProvider.NewFromCluster(r.Client, r.KafkaCluster)
-	if err != nil {
-		log.Info("could not create Kafka client, thus could not determine controller")
-		return brokers
-	}
-	defer close()
-
-	_, controllerID, err := kClient.DescribeCluster()
-	if err != nil {
-		log.Info("could not find controller broker")
-		return brokers
+// reorderBrokers returns the KafkaCluster brokers list reordered for reconciliation such that:
+// - the controller broker is reconciled last
+// - prioritize upscale in order to allow upscaling the cluster even when there is a stuck RU
+// - prioritize missing broker pods to be able for escaping from offline partitions, not all replicas in sync which
+//		could stall RU flow
+func reorderBrokers(brokerPods corev1.PodList, desiredBrokers []v1beta1.Broker, brokersState map[string]v1beta1.BrokerState, controllerBrokerID int32) []v1beta1.Broker {
+	runningBrokers := make(map[string]struct{})
+	for _, b := range brokerPods.Items {
+		brokerID := b.GetLabels()["brokerId"]
+		runningBrokers[brokerID] = struct{}{}
 	}
 
-	var controllerBroker *v1beta1.Broker
-	reorderedBrokers := make([]v1beta1.Broker, 0, len(brokers))
-	for i := range brokers {
-		if brokers[i].Id == controllerID {
-			controllerBroker = &brokers[i]
-		} else {
-			reorderedBrokers = append(reorderedBrokers, brokers[i])
+	// weights for broker ordering:
+	// -2: new broker
+	// -1: missing broker
+	// 0 : non-controller broker
+	// 1 : controller broker
+	brokersWeight := make(map[string]int, len(desiredBrokers))
+
+	for _, b := range desiredBrokers {
+		brokerID := fmt.Sprintf("%d", b.Id)
+		brokersWeight[brokerID] = 0
+
+		if _, ok := brokersState[brokerID]; !ok {
+			brokersWeight[brokerID] = -2 // new broker
```

Comment:
What do you think about creating an int enum to be more transparent about what a certain number means?